### DISCUSSION TITLE
Memory leaks discussion

### DISCUSSION THREAD

**Discussion by savelichalex** [software-mansion/react-native-reanimated#2286] [Q&A]
Hi guys. I've experienced huge memory leaks with reanimated@2.2.0 , I've been digging into a problem for almost a week, and still can't find a root cause.

So, I've tried to locate the problem with a simple component, but it seems that the problem persist in more complicated ones. (I even draw [a huge memory graph](https://app.diagrams.net/#G1TAALgnhlHRRzS-Vvp4Es1fz0hlzRE0-n) to identify loops, but for simple case it all looks safe and sound).

I've been looking into the problem with Instruments, and I want to share my measures.
This is a mount of one of our complicated components:
<img width="1740" alt="Screenshot 2021-08-13 at 23 27 12" src="https://user-images.githubusercontent.com/6286454/129414962-67f96b98-44a2-4124-9fbf-223e587a3b5a.png">

And this is what happens after unmount:
<img width="1740" alt="Screenshot 2021-08-13 at 23 28 30" src="https://user-images.githubusercontent.com/6286454/129415070-1a29470a-a8e9-4f83-8e2c-76f0fb1a51dc.png">

I've read about GC collection [there](https://github.com/software-mansion/react-native-reanimated/pull/1556) and I personally was able to catch the point of GCing things with simple debugger breakpoint in ShereableValue deconstuctor, so I know that it's necessary to wait a minute or two, so there another snapshot:
<img width="1740" alt="Screenshot 2021-08-13 at 23 31 38" src="https://user-images.githubusercontent.com/6286454/129415389-e3005cf1-8a5c-47e2-80c9-8493d3ad5c9a.png">

Moreover if I start to mount/unmount the same component, it seems that engine stops collection garbage from previous renders.

What worries me on this snapshots is this:
<img width="1740" alt="Screenshot 2021-08-13 at 23 34 03" src="https://user-images.githubusercontent.com/6286454/129415796-3431cdf8-8ba0-4f7c-8a92-07a5d1d4bb39.png">
If I understand it correctly it's either [worklets](https://github.com/software-mansion/react-native-reanimated/blob/master/Common/cpp/SharedItems/ShareableValue.cpp#L373) or [host functions](https://github.com/software-mansion/react-native-reanimated/blob/master/Common/cpp/SharedItems/ShareableValue.cpp#L324). What I can't understand is why there are so many of them. So it was a capture of one mount/unmount cycle, and I have around 30 `useAnimatedStyle` and few `useAnimatedReaction` in component. I can hardly imagine that it will produce more than 1k of refs.

Do you guys have any ideas? Maybe I miss something? 🙏

**Comment by terrysahaidak** [comment]
could you try latest alpha release to reproduce this?

**Reply by savelichalex** [reply]
Yes, I can, but I personally don't expect much difference as the only things were added are omptimization levels :)

**Reply by savelichalex** [reply]
I'll try it next week and share results

**Reply by savelichalex** [reply]
Tried it with reanimated@2.3.0-alpha.2, got the same results:
<img width="1740" alt="Screenshot 2021-08-15 at 17 14 23" src="https://user-images.githubusercontent.com/6286454/129481410-833cc0f4-b331-4c1f-a444-6e588d32f434.png">

**Reply by savelichalex** [reply]
Also I thought a bit, and did another measure, that time I've measured the second mount/unmount, as for the second pass all global worklets should already initialize cache, so it should be "pure" as far as I understand it.
(Please note that numbers are not related to previous ones, as I commented out part of the component, to simplify counting, more about it later).
<img width="1740" alt="Screenshot 2021-08-15 at 17 52 44" src="https://user-images.githubusercontent.com/6286454/129482578-c34f148a-3715-4947-b90b-d99b40e70c32.png">

This time I use 8 `useSharedValue` and 13 `useAnimatedStyle` (I had `useAnimatedGestureHandler` calls before, but I commented it out for now).
So If I count it right, for each `useAnimatedStyle` we have 3 new worklets (one passed in our code, and 2 is created in `useAnimatedStyle` hook, for a mapper). So it should be `13 * 3 = 39` new worklets. If I understand c++ name creation for lambdas correctly then `toJSValue(facebook::jsi::Runtime)::$_2` is [a worklet](https://github.com/software-mansion/react-native-reanimated/blob/master/Common/cpp/SharedItems/ShareableValue.cpp#L373), and it should be 1:1 mapping, but instead of 39 we see 2644, and 634 of them are not destroyed. I don't know how to properly interpretate this, to me personally it doesn't make any sense actually :)

**Reply by terrysahaidak** [reply]
I guess it's time for @Szymon20000 and @piaskowyk to take a look and verify everything.

**Reply by savelichalex** [reply]
Also another observation:
<img width="1740" alt="Screenshot 2021-08-15 at 18 04 36" src="https://user-images.githubusercontent.com/6286454/129483030-314f1107-c6fc-4e55-9973-94d37e657995.png">

It was created exactly 13 mappers (as it should), but `startMapper` function was created 39 times, that is 3 times more then it should be, now I'm curious how that happened 🤔

**Reply by savelichalex** [reply]
And another thing, is that `$_1` function seems to be created when there is a regular (so called `HostFunction`) is captured in a worklet, but I'm using only `interpolate` and `interpolateColor` in `useAnimatedStyle`

**Reply by savelichalex** [reply]
I was able to identify what `HostFunction`s have been called, it's [pushFrame](https://github.com/software-mansion/react-native-reanimated/blob/1d698d83c6f041603d548bf10d47eab992e50840/src/reanimated2/core.ts#L95) and [interpolateNode](https://github.com/software-mansion/react-native-reanimated/blob/1d698d83c6f041603d548bf10d47eab992e50840/src/reanimated2/interpolation.ts#L126) and it looks like they can be easily dropped for native, to not create unnecessary `ShareableValue`s

**Reply by savelichalex** [reply]
I've commented them out and got this picture:
<img width="1740" alt="Screenshot 2021-08-15 at 20 11 41" src="https://user-images.githubusercontent.com/6286454/129486644-4ddb710d-7daa-4578-98bf-e99e06232a51.png">

Definitely now there are less `SharedValue`s.

Interesting thing with worklets, looks like they were recreated 4 times, and one time they're stucked somewhere. All four times it's happened on mount.

**Reply by savelichalex** [reply]
To surpass a possible question - no additional dependencies aren't provided for `useAnimatedStyle`s. Also I tried to put `[]` as a dependency to have only `__workletHash` as [a dependency](https://github.com/software-mansion/react-native-reanimated/blob/master/src/reanimated2/Hooks.ts#L561), that in theory shouldn't trigger `useEffect`, but that didn't make any difference, there're still 3 times more `startMapper` calls than expected.

**Reply by savelichalex** [reply]
So, it can be reduced to just simple `useAnimatedStyle`, it doesn't look that significant that way, but it still contains the same leaked object as before:
<img width="1740" alt="Screenshot 2021-08-15 at 23 11 25" src="https://user-images.githubusercontent.com/6286454/129491385-2ae2f469-5caf-4ae1-9bb4-25d7294bd313.png">

**Reply by astanb** [reply]
Interesting stuff. I've found gesture handler and reanimated to be the cause of severe memory problem in our production app. I'm surprised there hasn't been a lot more discussion about the memory issues with these libraries, but maybe the majority of consumers are not using the apps that use these libraries for extended periods so the memory increase is not as noticeable. Our app is in constant use for long periods (3 hours +) and we've been having to tell our users to force close it to free up the memory. We are in the middle of a rewrite to remove all dependencies on gesture handler and reanimated, although https://github.com/software-mansion/react-native-reanimated/pull/3722 will hopefully get rid of these issues.

**Reply by terrysahaidak** [reply]
there is a PR in progress where the team is rewriting shared values to address all the memory leaks and performance problems with them.

**Reply by astanb** [reply]
Yep, its the PR I linked to in my post 👍
